README.md
# KrakenExchangeAPI-PowerShell

| :exclamation:  Module was moved to <https://github.com/voytas75/KrakenExchange> |
|---------------------------------------------------------------------------------|

[![PowerShell Gallery Version](https://img.shields.io/powershellgallery/v/krakenexchange?label=KrakenExchange%20)](https://www.powershellgallery.com/packages/KrakenExchange/2.0.0)

Note: The module and scripts are not affiliated with Kraken or any other cryptocurrency exchange, and use of is at your own risk.

This PowerShell scripts and module [KrakenExchange](https://www.powershellgallery.com/packages/KrakenExchange) provide an easy way to interact with the Kraken Exchange API. It includes cmdlets for querying account information, getting ticker information for a specific currency pair, ...

[KrakenExchange](https://www.powershellgallery.com/packages/KrakenExchange) is a PowerShell module that allows you to interact with the Kraken cryptocurrency exchange API. In this first version of the module, you can easily retrieve your account balance.

Future versions of the module will expand on this functionality, providing even more powerful tools for automating trading strategies and performing market analysis. You will be able to get ticker information for a specific asset pair, place buy and sell orders, get your trade history, and much more. The module also supports getting open orders and canceling orders, making it a comprehensive tool for managing your trades.

KrakenExchange is designed to be flexible and customizable, allowing you to easily integrate it into your existing PowerShell scripts and automation workflows. Whether you're a seasoned trader or just getting started with cryptocurrency trading, this module provides a powerful toolset to help you succeed in the market.

## Installation

### Module KrakenExchange

KrakenExchange 2.0.0 module has already arrived


Note: This PowerShell module is only compatible with **PowerShell version 7 and above**. If you are running an older version of PowerShell, you will need to upgrade to use this module.

Install [KrakenExchange](https://www.powershellgallery.com/packages/KrakenExchange) module or download scripts from [releases](https://github.com/voytas75/KrakenExchangeAPI-PowerShell/releases).

Installation:

```powershell
Install-Module -Name KrakenExchange
```

## ApiKey and ApiSecret

To get a Kraken API Key and API Secret, you will need to follow these steps:

1. Create a Kraken account if you don't already have one. You can sign up at <https://www.kraken.com/signup>.

2. Log in to your Kraken account and click on the "Settings" tab in the top right corner.

3. Select the "API" option from the dropdown menu.

4. Click the "Generate New Key" button.

5. You will need to give your API key a name and select the permissions you want to grant to the key. Kraken provides a variety of permission options, such as trading, funding, and viewing account balances.

6. Once you have selected your permissions, click the "Generate Key" button.

7. After generating your key, you will see your API Key and API Secret displayed on the screen. Make sure to copy both the API Key and API Secret to a safe location, as the API Secret will only be displayed once.

Note that it is important to keep your API Secret secure and not share it with anyone. You will need both your API Key and API Secret to access your Kraken account via API.

## KrakenExchange module usage

Function `Connect-KExchange` to provide API data for private access:

```powershell
Connect-KExchange
```

Function `Get-KEAssetInfo`:

```powershell
PS C:\> Get-KEAssetInfo -Asset eth | convertto-json
{
  "error": [],
  "result": {
    "XETH": {
...
    }
  }
}
```

Function `Get-KETicker` is using Websockets API and default `-Pair` is `BTC/USD`:

```powershell
PS C:\> Get-KETicker

a : {27848.30000, 0, 0.10000000}
...
o : {28179.00000, 28496.40000}
```

[message ticker](https://docs.kraken.com/websockets/#message-ticker)

Function `Get-KETickerInformation`:

```powershell
PS C:\> Get-KETickerInformation -Pair "BTCUSD" | ConvertTo-Json -Depth 3
{
  "error": [],
  "result": {
    "XXBTZUSD": {
      "a": [
        "27911.10000",
...
      "o": "28179.00000"
    }
  }
}
```

Function `Get-KETradableAssetPair`:

```powershell
PS C:\> Get-KETradableAssetPair -Pair "ethusd" -Info info | ConvertTo-Json -Depth 4
{
  "error": [],
  "result": {
    "XETHZUSD": {
      "altname": "ETHUSD",
...
      "short_position_limit": 1800
    }
  }
}
```

Function `Get-KEAssetInfo`:

```powershell
PS C:\> Get-KEAssetInfo -Asset "eth,btc" | ConvertTo-Json
{
  "error": [],
  "result": {
    "XETH": {
...
    },
    "XXBT": {
...
    }
  }
}
```

Function `Disconnect-KExchange` to clear API data:

```powershell
Disconnect-KExchange
```

## Scripts usage

Here are some examples of how to use scripts. The most important is the REST API client for public and private environments:

```powershell
# Get trade volume using wrapper
$env:ApiKey = "<Api_Key>"
$env:ApiSecret = "<Api_Secret>"
$Result = (.\kraken.ps1 -ApiKey $env:apikey -ApiSecret $env:ApiSecret -TradeVolume).result 
$Result | Select-Object -ExpandProperty volume
```

Other examples:

```powershell
# Get ticker information for BTC/USD
.\Get-KrakenTicker.ps1 -Pair 'ETH/USD'

# Get signature
PS C:\> $payload = [System.Collections.Specialized.OrderedDictionary]::new()
PS C:\> $payload.Add("nonce", [Math]::Round((New-TimeSpan -Start "1/1/1970").TotalMilliseconds))
PS C:\> $payload.Add("ordertype", "limit")
PS C:\> $payload.Add("type", "buy")
PS C:\> $payload.Add("pair", "XXBTZUSD")
PS C:\> $payload.Add("price", "9000")
PS C:\> $payload.Add("volume", "0.01")
PS C:\> .\Set-APIKrakenSignature.ps1 -Payload $payload -URI "/0/private/AddOrder" -api_secret "KrakenAPIsecret"
```

## Versioning

We use [SemVer](http://semver.org/) for versioning.

## Contributing

Contributions to this project are welcome and encouraged! If you notice a bug, have an idea for a new feature, or would like to contribute code, please open a GitHub issue or pull request.

Before contributing, please review our [contribution guidelines](CONTRIBUTING.md) for instructions on how to get started.

## License

This project is licensed under the terms of the [GNU General Public License (GPL) version 3](LICENSE).

The GPL is a copyleft license, which means that any derivative works must be distributed under the same terms. This ensures that the code remains free and open source.

For more information about the GPL, please see the [official GNU website](https://www.gnu.org/licenses/gpl-3.0.en.html).

## Contact

If you have any questions or comments about this project, please feel free to [contact us](mailto:wnapierala@hotmail.com). We would be happy to hear from you!


KrakenExchange/ReleaseNotes.md
# `KrakenExchange` powershell module - release notes

## 2.1.0 - 2023.04.30

### Added

- param alias in `Find-KEZeroProfitPrice`,
- new function `Get-KEClosedOrders`,
- new function `Get-KEOrdersInfo`,
- new function `Get-KETradesHistory`,
- new function `Get-KETradesInfo`,
- new function `Get-KEOpenPositions`,
- alias `Get-KEAsset` for function `Get-KEAssetInfo`,
- alias `Get-KEAssets` for function `Get-KEAssetInfo`,
- alias `Get-KEPairs` for function `Get-KETradableAssetPair`,
- alias `Get-KEAssetPairs` for function `Get-KETradableAssetPair`,
- interval `21600` to `Get-KEOHLCData`,
- new private function `New-KEDataFolder` for create kraken data folder.

### Changed

- output of `Find-KEProfit`,
- validate set fo param `OHLCInterval`,
- show all assets by function `Get-KEAssetInfo`,
- show all pairs by function `Get-KETradableAssetPair`,
- `validatepattern` for asset, crypto and currency.

## 2.0.0 - 2023.04.09

### Added

- new function `Connect-KExchange`
- new function `Disconnect-KExchange`
- new function `Get-KETradeBalance`
- new function `Get-KEOpenOrders`
- new function `Get-KETradeVolume`

### Changed

- secure api key and secret,
- minor changes.

## 1.1.0 - 2023.04.06

### Added

- added `Get-KETradeVolume` function to retrieve trading volume information for a specified asset pair.
- added `Get-KEOHLCData` function to retrieve Open, High, Low, and Close data for a specified asset pair.
- added `Get-KEOrderBook` to retrieve the order book for a given trading pair.
- added `Get-KERecentSpreads` to retrieve recent spreads for a specified trading pair.
- added `Get-KERecentTrades` to retrieve recent trades for a specified trading pair.

### Changed

- secure api key and secret,
- fixed issue with `Get-KETickerInformation` function not returning correct ticker information for certain asset pairs,
- updated help documentation for all functions to provide additional information and usage examples,
- refactored code for improved performance and maintainability,
- minor changes.

## 1.0.1 - 2023.04.04

### Added

- new functions,
- secure api key and secret.

### Changed

- minor changes.

## 1.0.0 - 2023.03

### Added

- functions for public endpoints.

### Changed

- minor changes.


Set-APIKrakenSignature.ps1
function Set-APIKrakenSignature {
    <#
.SYNOPSIS
    Calculates the signature for a Kraken API request.

.DESCRIPTION
    Calculates the signature for a Kraken API request using the specified payload, URI, and API secret.

.PARAMETER Payload
    The payload for the API request, as an OrderedDictionary.

.PARAMETER URI
    The URI for the API request, as a string.

.PARAMETER api_secret
    The API secret for the Kraken account, as a string.
.NOTES
    Both [System.Security.Cryptography.HashAlgorithm]::Create('SHA256') and [System.Security.Cryptography.SHA256]::Create() create an instance of the SHA256 hashing algorithm provided by the .NET Framework's System.Security.Cryptography namespace. The difference between them is that the former creates an instance of the HashAlgorithm class and the latter creates an instance of the SHA256 class, which inherits from HashAlgorithm.
    The HashAlgorithm class is an abstract base class that provides the basic functionality common to all hash algorithms. It defines the ComputeHash method, which computes the hash value for a given input, as well as other methods and properties that are common to all hash algorithms. When you create an instance of the SHA256 algorithm using [System.Security.Cryptography.HashAlgorithm]::Create('SHA256'), you are creating an instance of the SHA256Managed class, which is a concrete implementation of the SHA256 algorithm that inherits from HashAlgorithm.
    On the other hand, when you create an instance of the SHA256 algorithm using [System.Security.Cryptography.SHA256]::Create(), you are creating an instance of the SHA256Cng class, which is a concrete implementation of the SHA256 algorithm that inherits directly from SHA256.
    In general, it is recommended to use the specific algorithm class, such as SHA256, rather than the HashAlgorithm base class, as it provides a more strongly-typed interface and may have additional functionality specific to that algorithm. However, both methods should work fine for creating an instance of the SHA256 algorithm.
    [System.Security.Cryptography.HMAC]::Create("HMACSHA512") - it works too.LINK
    Specify a URI to a help page, this will show when Get-Help -Online is used.
.EXAMPLE
    PS C:\> $payload = [System.Collections.Specialized.OrderedDictionary]::new()
    PS C:\> $payload.Add("nonce", [int64]([DateTime]::UtcNow - (New-Object DateTime 1970, 1, 1, 0, 0, 0, 0, ([DateTimeKind]::Utc))).TotalMilliseconds
    PS C:\> $payload.Add("ordertype", "limit")
    PS C:\> $payload.Add("type", "buy")
    PS C:\> $payload.Add("pair", "XXBTZUSD")
    PS C:\> $payload.Add("price", "9000")
    PS C:\> $payload.Add("volume", "0.01")
    PS C:\> Set-APIKrakenSignature -Payload $payload -URI "/0/private/AddOrder" -api_secret "KrakenAPIsecret"

.NOTES
    Author: wnapierala [@] hotmail.com
    Date: 03.2023
#>

    [CmdletBinding()]
    param (
        [Parameter(Mandatory = $true, HelpMessage = "The payload for the API request, as an OrderedDictionary.")]
        [System.Collections.Specialized.OrderedDictionary]$Payload,

        [Parameter(Mandatory = $true, HelpMessage = "The URI for the API request, as a string.")]
        [string]$URI,

        [Parameter(Mandatory = $true, HelpMessage = "The API secret for the Kraken account, as a string.")]
        [string]$api_secret
    )

    # Convert the payload to a URL-encoded string
    $url_encoded_payload = ($payload.GetEnumerator() | ForEach-Object { $_.Name + "=" + $_.Value }) -join "&"

    # Create an instance of the SHA256 hashing algorithm
    $sha = [System.Security.Cryptography.SHA256]::Create()

    # Compute the SHA256 hash of the nonce and URL-encoded payload
    $shahash = $sha.ComputeHash([System.Text.Encoding]::UTF8.GetBytes($Payload["nonce"].ToString() + $url_encoded_payload))

    # Create an instance of the HMAC-SHA512 algorithm
    $mac = New-Object System.Security.Cryptography.HMACSHA512

    # Set the key to the API secret converted to bytes
    $api_secret_bytes = [System.Convert]::FromBase64String($api_secret)
    $mac.Key = $api_secret_bytes

    # Compute the HMAC-SHA512 hash of the URI and SHA256 hash
    $machash = $mac.ComputeHash([System.Text.Encoding]::UTF8.GetBytes($URI) + $shahash)

    # Convert the result to a Base64-encoded string and return it
    $signature = [System.Convert]::ToBase64String($machash)
    return $signature
}
